rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,160 @@
+
+Recent work on the AWS `codebuild`, `codedeploy`, and `codepipeline` services has demonstrated that only "Two-Way" mode is actually viable: `codebuild`, for instance, accepts an arbitrary, user-defined YAML file describing the process to build code from some place and store it in some other place. IaSQL cannot know before it triggers `codebuild` what it will do, and so during the `iasql_apply` call suddenly there will exist resources, such as a new ECR image, that are not in the database which `iasql_apply` will want to remove because that's not part of the "desired" state as far as the engine is concerned, but *is* the true desired outcome from the user.
+
+Attempting to `iasql_sync` in the middle of that `iasql_apply` could cause IaSQL to drop other operations it is supposed to be performing, freezing their cloud into some unexpected intermediate state. And it is impossible to "parse" the YAML to determine what resources are going to be created. Since the YAML allows arbitrary shell script execution, the script could do something like `if (Math.random() > 0.5) { writeToS3(); } else { writeToECR(); }` or even `curl https://raw.githubusercontent.com/company/project/some_binary_that_calls_aws_apis; chmod +x some_binary_that_calls_aws_apis; ./some_binary_that_calls_aws_apis;` and there's no good, general purpose way to know what the side-effects of `codebuild` will be.
+
+However, it is possible for "Two-Way" mode to emulate the other three in a way that is also mostly-backwards-compatible with the current manual mode. Furthermore, a singular IaSQL database would be able to have different "parts" of the database emulating different modes at the same time.
+
+To accomplish this, let's assume under normal operations, there is a singular apply/sync-combined command to perform the "Two-Way"-style action. It works similarly to the current sync and apply, but also uses the audit log to decide for each record which behavior to follow. This piece can be operated in a "manual" style, invoked explicitly, but we'll instead invoke it through a cron-like interval, every few minutes, to make the automatic mode. In this way, there is no special syntax for users to work with normally. They just `SELECT/INSERT/UPDATE/DELETE` records they see and eventually this is reflected in their cloud account.
+
+To emulate manual mode would basically be to temporarily turn off that cron, let the user make whatever changes they want to batch together, and then re-enable that cron. Because we can't actually support the current `iasql_apply` for the three `code-` services, and because entering this batching mode that could potentially be unwound, it looks a *lot* like a transaction, so the proposed new syntax would be to emulate that:
+
+```sql
```

Comment:
is there a way to make it atomic? or , if we detect that the transaction has stopped in the middle for x reason , rollback to a safe state ?

Reply:
I think it should be possible via a lock (it can simply be a column in the metadata table akin to what we do for updates) that the cron checks

Reply:
David mentions it here https://github.com/iasql/iasql-engine/pull/1432/files#r1003607851